src/utils.js

```diff
@@ -71,37 +72,65 @@ const _runJavascript = (jsString, window, filename = 'script', lineOffset = 0, c
 };
 module.exports._runJavascript = _runJavascript;
 
-const _normalizeBuffer = (b, target) => {
-  const name = b && b.constructor && b.constructor.name;
-  switch (name) {
-    case 'Buffer':
-    case 'ArrayBuffer':
-    case 'Uint8Array':
-    case 'Uint8ClampedArray':
-    case 'Int8Array':
-    case 'Uint16Array':
-    case 'Int16Array':
-    case 'Uint32Array':
-    case 'Int32Array':
-    case 'Float32Array':
-    case 'Float64Array':
-    case 'DataView':
-    case 'Promise': {
-      if (!(b instanceof target[name])) {
-        GlobalContext.nativeVm.setPrototype(b, target[name].prototype);
-      }
-      break;
-    }
-    case 'Blob': {
-      if (!(b.buffer instanceof target.Buffer)) {
-        GlobalContext.nativeVm.setPrototype(b.buffer, target.Buffer.prototype);
-      }
-      break;
-    }
+/**
+ * Normalize prototype between native and V8.
+ * Node's vm module was slow so there's vm-one for one-way bindings:
+ *   github.com/modulesio/vm-one
+ * nativeVm.setPrototype will do the nan casting setPrototype between types.
+ * Required for instanceof to work.
+ *
+ * @param obj - Object to convert.
+ * @param targetContext - window or vm context that prototype will get retrieved from.
+ */
+const _normalizePrototype = (obj, targetContext) => {
+  let name = obj && obj.constructor && obj.constructor.name;
+  if (obj instanceof targetContext[name]) { return obj; }
+
+  const isToWindow = !!targetContext[symbols.prototypesSymbol];
+
+  // Convert Blob's buffer.
+  if (name === 'Blob') {
+    _normalizePrototype(obj.buffer, targetContext);
+    return obj;
   }
-  return b;
+
+  // Normalize to window prototype.
+  if (isToWindow) {
+    GlobalContext.nativeVm.setPrototype(obj, targetContext[symbols.prototypesSymbol][name] ||
+                                             targetContext[name].prototype);
+    return obj;
+  }
+
+  // Normalize to native prototype.
+  GlobalContext.nativeVm.setPrototype(obj, targetContext[name].prototype);
```

Comment:
Seems we can fall though here if `obj` is something unexpected, like `null`. I don't know if trying to set the prototype will explode in that case, but I _do_ think callers are expecting this function to be safe for all inputs.

The original code was fine since it wouldn't [pass the `name` check](https://github.com/webmixedreality/exokit/pull/441/files#diff-2b4ca49d4bb0a774c4d4c1672d7aa781L76).

We should either confirm that this is fine for all inputs (hard) or be safe and do e.g. as prelude to this function

```
if (obj && typeof obj === 'object') { // non-null object
```

We also would probably be guaranteed to have a `constructor.name` in that case.

Reply:
Done, added unit tests.